### PR TITLE
Add more cases supported by 'use collection expression'

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/75870

### CODE REVIEW

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForCreateDiagnosticAnalyzer.cs [hunk 1]
```diff
@@ -20,11 +20,6 @@ internal sealed partial class CSharpUseCollectionExpressionForCreateDiagnosticAn
         IDEDiagnosticIds.UseCollectionExpressionForCreateDiagnosticId,
         EnforceOnBuildValues.UseCollectionExpressionForCreate)
 {
-    public const string UnwrapArgument = nameof(UnwrapArgument);
-
-    private static readonly ImmutableDictionary<string, string?> s_unwrapArgumentProperties =
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForCreateDiagnosticAnalyzer.cs:25]: moved to common location

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForCreateDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -52,9 +47,7 @@ private void AnalyzeInvocationExpression(SyntaxNodeAnalysisContext context, INam
         }
 
         var locations = ImmutableArray.Create(invocationExpression.GetLocation());
-        var properties = unwrapArgument ? s_unwrapArgumentProperties : ImmutableDictionary<string, string?>.Empty;
-        if (changesSemantics)
-            properties = properties.Add(UseCollectionInitializerHelpers.ChangesSemanticsName, "");
+        var properties = GetDiagnosticProperties(unwrapArgument, useSpread, changesSemantics);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForCreateDiagnosticAnalyzer.cs:50]: moved to common location

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -63,17 +63,6 @@ internal sealed partial class CSharpUseCollectionExpressionForFluentDiagnosticAn
...
-    /// <summary>
-    /// Set of type-names that are blocked from moving over to collection expressions because the semantics of them are
-    /// known to be specialized, and thus could change semantics in undesirable ways if the compiler emitted its own
-    /// code as an replacement.
-    /// </summary>
-    private static readonly ImmutableHashSet<string?> s_bannedTypes = [
-        nameof(ParallelEnumerable),
-        nameof(ParallelQuery),
-        // Special internal runtime interface that is optimized for fast path conversions of collections.
-        "IIListProvider"];
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer.cs:75]: moved to common location

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -341,36 +330,8 @@ bool IsListLike(ExpressionSyntax expression)
                 return false;
 
             return
-                Implements(type, compilation.IListOfTType()) ||
-                Implements(type, compilation.IReadOnlyListOfTType());
-        }
-
-        bool IsIterable(ExpressionSyntax expression)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForFluentDiagnosticAnalyzer.cs:348]: moved to common location

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForNewDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+namespace Microsoft.CodeAnalysis.CSharp.UseCollectionExpression;
+
+using static UseCollectionExpressionHelpers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+internal sealed partial class CSharpUseCollectionExpressionForNewDiagnosticAnalyzer()
+    : AbstractCSharpUseCollectionExpressionDiagnosticAnalyzer(
+        IDEDiagnosticIds.UseCollectionExpressionForNewDiagnosticId,
+        EnforceOnBuildValues.UseCollectionExpressionForNew)
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/CSharpUseCollectionExpressionForNewDiagnosticAnalyzer.cs:23]: supports converting `new List<...>(x)` to `[.. x]`

File: src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs [hunk 6]
```diff
@@ -1016,30 +1035,11 @@ bool IsCompatibleSignatureAndArguments(
                             Name: nameof(IEnumerable<int>),
                             TypeArguments: [ITypeParameterSymbol { TypeParameterKind: TypeParameterKind.Method }]
                         } enumerableType
-                    }] && enumerableType.OriginalDefinition.Equals(compilation.IEnumerableOfTType()))
+                    }] &&
+                    enumerableType.OriginalDefinition.Equals(compilation.IEnumerableOfTType()) &&
+                    arguments.Count == 1)
                 {
-                    var argExpression = arguments[0].Expression;
-                    if (argExpression
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Analyzers/UseCollectionExpression/UseCollectionExpressionHelpers.cs:1022]: moved to helper.

File: src/Analyzers/CSharp/Tests/CSharpAnalyzers.UnitTests.projitems [hunk 7]
```diff
@@ -115,6 +115,7 @@
     <Compile Include="$(MSBuildThisFileDirectory)UseCoalesceExpression\UseCoalesceExpressionForTernaryConditionalCheckTests.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\UseCollectionExpressionForFluentTests.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\UseCollectionExpressionForCreateTests.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)usecollectionexpression\UseCollectionExpressionForNewTests.cs" />
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/Tests/CSharpAnalyzers.UnitTests.projitems:118]:
```suggestion
    <Compile Include="$(MSBuildThisFileDirectory)UseCollectionExpression\UseCollectionExpressionForNewTests.cs" />
```

**Comment by CyrusNajmabadi** [comment]
@akhera99 @dibarbet ptal.

**Review by akhera99 — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForFluentTests.cs [hunk 8]
```diff
@@ -2711,6 +2711,55 @@ public async Task TestNotInParallelEnumerable()
...
+                        var result = sequence.AsParallel().ToArray();
+                    }
+                }
+                """,
+            LanguageVersion = LanguageVersion.CSharp12,
+        }.RunAsync();
+    }
+
+    [Fact, WorkItem("https://github.com/dotnet/roslyn/issues/71145")]
+    public async Task TestNotInParallelEnumerable2()
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Analyzers/CSharp/Tests/UseCollectionExpression/UseCollectionExpressionForFluentTests.cs:2731]: Correct.  We didn't update hte parallel one, only the non parallel one :)

**Pull request merged**